File: core/src/main/java/com/lomekwi/cave/pipeline/text/TextFrame.java
package com.lomekwi.cave.pipeline.text;

import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.graphics.g2d.BitmapFont;
import com.badlogic.gdx.graphics.g2d.GlyphLayout;
import com.badlogic.gdx.math.Matrix4;
import com.lomekwi.cave.pipeline.Frame;
import com.lomekwi.cave.pipeline.Source;
import com.lomekwi.cave.pipeline.image.Transform;
import com.lomekwi.cave.pipeline.image.Transformable;
import com.lomekwi.cave.timeline.Track;
import com.lomekwi.cave.ui.editpanel.previewarea.TransFrameActor;

public class TextFrame extends Frame implements Transformable {
    private volatile String text;
    private volatile BitmapFont font;
    private final GlyphLayout layout = new GlyphLayout();
    private Transform transform;
    private TransFrameActor actor;
    private final Matrix4 tmpMatrix = new Matrix4();
    private volatile boolean glyphsMissing;
    private volatile float cachedWidth;
    private volatile float cachedHeight;
    private volatile int version;
    private int layoutVersion;

    public TextFrame(Track track) {
        super(track);
    }

    public TextFrame(Track track, Source<?> source) {
        super(track, source);
    }

    public void setText(CharSequence text) {
        this.text = text.toString();
        version++;
    }

    public void setFont(BitmapFont font) {
        this.font = font;
        version++;
    }

    public void initActor() {
        actor = new TransFrameActor(this);
    }

    public TransFrameActor getActor() {
        return actor;
    }

    @Override
    public Transform getTransform() {
        return transform;
    }

    @Override
    public void setTransform(Transform transform) {
        this.transform = transform;
    }

    @Override
    public float getBaseWidth() {
        return cachedWidth;
    }

    @Override
    public float getBaseHeight() {
        return cachedHeight;
    }

    @Override
    public void render(Batch batch) {
        if (version != layoutVersion) {
            rebuildLayout();
            layoutVersion = version;
        }
        if (font == null || text == null || glyphsMissing) return;
        var t = getTransform();
        float scaleX = t.flipX ? -1 : 1;
        float scaleY = t.flipY ? -1 : 1;
        float w = cachedWidth;
        float h = cachedHeight;

        Matrix4 saved = new Matrix4(batch.getTransformMatrix());
        float sx = scaleX * t.getScaleX();
        float sy = scaleY * t.getScaleY();
        tmpMatrix.set(saved);
        tmpMatrix.translate(t.x + w * sx / 2, t.y + h * sy / 2, 0);
        tmpMatrix.rotate(0, 0, 1, t.rotation);
        tmpMatrix.scale(sx, sy, 1);
        batch.setTransformMatrix(tmpMatrix);
        try {
            font.draw(batch, layout, -w / 2, -font.getDescent());
        } catch (NullPointerException e) {
            glyphsMissing = true;
        }
        batch.setTransformMatrix(saved);
    }

    private void rebuildLayout() {
        String t = text;
        BitmapFont f = font;
        if (f == null || t == null) {
            glyphsMissing = true;
            cachedWidth = 0;
            cachedHeight = 0;
            return;
        }
        for (int i = 0; i < t.length(); i++) {
            if (f.getData().getGlyph(t.charAt(i)) == null) {
                glyphsMissing = true;
                cachedWidth = 0;
                cachedHeight = 0;
                return;
            }
        }
        layout.setText(f, t);
        glyphsMissing = false;
        cachedWidth = layout.width;
        cachedHeight = layout.height;
    }
}

File: core/src/main/scala/com/lomekwi/cave/pipeline/text/TextFrame.scala
package com.lomekwi.cave.pipeline.text

import com.badlogic.gdx.graphics.Color.WHITE

import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.graphics.g2d.BitmapFont
import com.badlogic.gdx.graphics.g2d.GlyphLayout
import com.badlogic.gdx.math.Matrix4
import com.lomekwi.cave.pipeline.Frame
import com.lomekwi.cave.pipeline.Segment
import com.lomekwi.cave.pipeline.image.Transform
import com.lomekwi.cave.pipeline.image.Transformable
import com.lomekwi.cave.timeline.Track
import com.lomekwi.cave.ui.editpanel.previewarea.TransFrameActor

import scala.compiletime.uninitialized
import scala.jdk.CollectionConverters.*

class TextFrame(trackIndex: Int, segment: Segment[?]) extends Frame(trackIndex, segment) with Transformable {
  @volatile private var text: String = uninitialized
  @volatile private var font: BitmapFont = uninitialized
  private final val layout: GlyphLayout = new GlyphLayout()
  private var transform: Transform = uninitialized
  private var actor: TransFrameActor = uninitialized
  private final val tmpMatrix: Matrix4 = new Matrix4()
  @volatile private var glyphsMissing: Boolean = false
  @volatile private var cachedWidth: Float = 0f
  @volatile private var cachedHeight: Float = 0f
  @volatile private var cachedCenterY: Float = 0f
  @volatile private var version: Int = 0
  private var layoutVersion: Int = 0

  def this(trackIndex: Int) = {
    this(trackIndex, null)
  }

  def setText(text: CharSequence): Unit = {
    this.text = text.toString
    version += 1
  }

  def setFont(font: BitmapFont): Unit = {
    this.font = font
    version += 1
  }

  def setActor(actor: TransFrameActor): Unit = {
    this.actor = actor
  }

  def getActor: TransFrameActor = {
    actor
  }

  override def getTransform: Transform = {
    transform
  }

  override def setTransform(transform: Transform): Unit = {
    this.transform = transform
  }

  override def getBaseWidth: Float = {
    cachedWidth
  }

  override def getBaseHeight: Float = {
    cachedHeight
  }

  override def render(batch: Batch): Unit = {
    if (version != layoutVersion) {
      rebuildLayout()
      layoutVersion = version
    }
    if (font != null && text != null && !glyphsMissing) {
      val t = getTransform
      val scaleX = if (t.isFlipX) -1f else 1f
      val scaleY = if (t.isFlipY) -1f else 1f
      val w = cachedWidth
      val h = cachedHeight

      val saved = new Matrix4(batch.getTransformMatrix)
      val sx = scaleX * t.getScaleX
      val sy = scaleY * t.getScaleY
      tmpMatrix.set(saved)
      tmpMatrix.translate(t.getX + w * sx / 2, t.getY + h * sy / 2, 0)
      tmpMatrix.rotate(0, 0, 1, t.getRotation)
      tmpMatrix.scale(sx, sy, 1)
      batch.setTransformMatrix(tmpMatrix)
      font.setColor(WHITE)
      try {
        font.draw(batch, layout, -w / 2, -cachedCenterY)
      } catch {
        case _: NullPointerException =>
          glyphsMissing = true
      }
      batch.setTransformMatrix(saved)
    }
  }

  private def rebuildLayout(): Unit = {
    val t = text
    val f = font
    if (f == null || t == null || hasMissingGlyph(f, t)) {
      glyphsMissing = true
      cachedWidth = 0
      cachedHeight = 0
      cachedCenterY = 0
    } else {
      font.setColor(WHITE)
      layout.setText(f, t)
      glyphsMissing = false
      cachedWidth = layout.width

      var minBottom = Float.MaxValue
      var maxTop = -Float.MaxValue
      for (run <- layout.runs.asScala) {
        for (glyph <- run.glyphs.asScala) {
          val bottom = run.y + glyph.yoffset
          val top = bottom + glyph.height
          if (bottom < minBottom) minBottom = bottom
          if (top > maxTop) maxTop = top
        }
      }
      if (minBottom > maxTop) {
        cachedHeight = layout.height
        cachedCenterY = font.getAscent
      } else {
        cachedHeight = maxTop - minBottom
        cachedCenterY = font.getAscent + (maxTop + minBottom) / 2f
      }
    }
  }

  /** 文本中含有字体缺少字形的非空白字符。 */
  private def hasMissingGlyph(f: BitmapFont, t: String): Boolean = {
    var i = 0
    var missing = false
    while (i < t.length() && !missing) {
      val c = t.charAt(i)
      if (!(c == ' ' || c == '\n' || c == '\r' || c == '\t')) {
        missing = f.getData.getGlyph(c) == null
      }
      i += 1
    }
    missing
  }
}
